internal/provider/resource_ko_resolve.go

```diff
@@ -0,0 +1,263 @@
+	opts, err := NewProviderOpts(meta)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &resolver{
+		bo: opts.bo,
+		po: opts.po,
+		fo: &options.FilenameOptions{},
+		so: &options.SelectorOptions{},
+	}
+
+	if p, ok := d.Get(BaseImageKey).(string); ok {
+		r.bo.BaseImage = p
+	}
+
+	if p, ok := d.Get(TagsKey).([]interface{}); ok {
+		if len(p) == 0 {
+			r.po.Tags = []string{"latest"}
+		} else {
+			r.po.Tags = StringSlice(p)
+		}
+	}
+
+	if p, ok := d.Get(PushKey).(bool); ok {
+		r.po.Push = p
+	}
+
+	if p, ok := d.Get(FilenamesKey).([]interface{}); ok {
+		r.fo.Filenames = StringSlice(p)
+	}
+
+	if p, ok := d.Get(RecursiveKey).(bool); ok {
+		r.fo.Recursive = p
+	}
+
+	if p, ok := d.Get(SelectorKey).(string); ok {
+		r.so.Selector = p
+	}
+
+	return r, nil
+}
+
+func (r *resolver) Resolve(ctx context.Context) (*Resolved, error) {
+	builder, err := commands.NewBuilder(ctx, r.bo)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheBuilder, err := build.NewCaching(builder)
+	if err != nil {
+		return nil, err
+	}
+
+	publisher, err := commands.NewPublisher(r.po)
+	if err != nil {
+		return nil, err
+	}
+	defer publisher.Close()
+
+	var resolveBuf bytes.Buffer
+	w := &nopWriteCloser{Writer: bufio.NewWriter(&resolveBuf)}
+
+	if err := commands.ResolveFilesToWriter(ctx, cacheBuilder, publisher, r.fo, r.so, w); err != nil {
+		return nil, err
+	}
+
+	if err := w.Flush(); err != nil {
+		return nil, err
+	}
+
+	// Split the dump of multi-doc yaml back into their individual nodes
+	// NOTE: Don't use a strings.Split to ensure we filter out any null docs
+	manifests := []string{}
+	decoder := yaml.NewDecoder(&resolveBuf)
+	for {
+		// Use an interface instead of yaml.Node to easily strip whitespaces and nil docs
+		var d interface{}
+		if err := decoder.Decode(&d); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, err
+		}
+
+		if d == nil {
+			continue
+		}
+
+		var buf bytes.Buffer
+		if err := yaml.NewEncoder(&buf).Encode(&d); err != nil {
+			return nil, err
+		}
+		manifests = append(manifests, buf.String())
+	}
+
+	sha256hash := sha256.Sum256(resolveBuf.Bytes())
+
+	return &Resolved{
+		Id:        fmt.Sprintf("%x", sha256hash),
+		Manifests: manifests,
+	}, nil
+}
+
+func resourceKoResolveCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	r, err := NewResolver(d, meta)
+	if err != nil {
+		return diag.Errorf("building resolver: %v", err)
+	}
+
+	resolved, err := r.Resolve(ctx)
+	if err != nil {
+		return diag.Errorf("resolving: %v", err)
+	}
+
+	d.SetId(resolved.Id)
+	d.Set("manifests", resolved.Manifests)
+	return nil
+}
+
+func resourceKoResolveRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	r, err := NewResolver(d, meta)
+	if err != nil {
+		return diag.Errorf("building resolver: %v", err)
+	}
+
+	// NOTE: Fake the publisher to prevent needing to rebuild the image on reads
```

Comment:
Does this actually prevent rebuilds? 🤔 

`ko_image` rebuilds on reads, but I don't know of any way to avoid that. We want there to be a TF diff if either YAML contents or Go source contents change, right?

Reply:
I think I messed my verbiage up, iiuc this prevents re-_publishing_, but builds still need to happen (for digests diffs) on reads.

that being said, I guessed on the best way to skip publishes, if there's a simpler way lmk!

Reply:
Yeah this will rebuild, but not republish, which I think is what we want / the best we can do.